pandas/lib.pyx

```diff
@@ -1055,6 +1055,20 @@ def string_array_replace_from_nan_rep(ndarray[object, ndim=1] arr, object nan_re
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
+def object_array_decode_bytes(ndarray[object, ndim=1] arr, object encoding):
+    """Decode any instances of bytes to str in arr using the given encoding."""
```

Comment:
@jreback I have just pushed a replacement commit here which moves most of it to `to_native_types()` as you requested (good idea by the way).  There remains one place that wasn't handled by that--namely the upper-left cell, the DataFrame.index.name, because that doesn't come from a series.  So I left bytes_to_str() for that one use.

I have also left the new Cython (expanded it actually, to handle MultiIndexes).  If what you are saying is that you prefer not to have code which handles cases of mixed bytes and str in a single column, I could deliberately alter it to avoid supporting that, but the code will still be needed to handle the MultiIndex case (unless you also do not want to support that).

What is here now I think addresses a lot of your original concerns, so please do review again.